subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/RoutingVirtualFileSystem.java

```diff
@@ -0,0 +1,102 @@
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class RoutingVirtualFileSystem implements VirtualFileSystem {
+    private final WellKnownFileLocations gradleUserHomeFileLocations;
+    private final VirtualFileSystem gradleUserHomeVirtualFileSystem;
+    private final VirtualFileSystem buildScopedVirtualFileSystem;
+
+    public RoutingVirtualFileSystem(
+        WellKnownFileLocations gradleUserHomeLocations,
+        VirtualFileSystem gradleUserHomeVirtualFileSystem,
+        VirtualFileSystem buildScopedVirtualFileSystem
+    ) {
+        this.gradleUserHomeFileLocations = gradleUserHomeLocations;
+        this.gradleUserHomeVirtualFileSystem = gradleUserHomeVirtualFileSystem;
+        this.buildScopedVirtualFileSystem = buildScopedVirtualFileSystem;
+    }
+
+    @Override
+    public void read(String location, FileSystemSnapshotVisitor visitor) {
+        getVirtualFileSystemFor(location).read(location, visitor);
+    }
+
+    @Override
+    public <T> Optional<T> readRegularFileContentHash(String location, Function<HashCode, T> visitor) {
+        return getVirtualFileSystemFor(location).readRegularFileContentHash(location, visitor);
+    }
+
+    @Override
+    public void read(String location, SnapshottingFilter filter, FileSystemSnapshotVisitor visitor) {
+        getVirtualFileSystemFor(location).read(location, filter, visitor);
+    }
+
+    @Override
+    public void update(Iterable<String> locations, Runnable action) {
+        if (Iterables.isEmpty(locations)) {
+            return;
+        }
+        Iterable<String> immutableLocations = Iterables.filter(locations, gradleUserHomeFileLocations::isImmutable);
+        int immutableLocationsSize = Iterables.size(immutableLocations);
+        if (immutableLocationsSize == 0) {
+            buildScopedVirtualFileSystem.update(locations, action);
+        } else if (immutableLocationsSize == Iterables.size(locations)) {
+            gradleUserHomeVirtualFileSystem.update(locations, action);
+        } else {
+            Iterable<String> mutableLocations = Iterables.filter(locations, location -> !gradleUserHomeFileLocations.isImmutable(location));
+            gradleUserHomeVirtualFileSystem.update(immutableLocations, action);
+            buildScopedVirtualFileSystem.update(mutableLocations, action);
+        }
+    }
+
+    @Override
+    public void invalidateAll() {
+        gradleUserHomeVirtualFileSystem.invalidateAll();
+        buildScopedVirtualFileSystem.invalidateAll();
+    }
+
+    @Override
+    public void updateWithKnownSnapshot(String location, FileSystemLocationSnapshot snapshot) {
+        getVirtualFileSystemFor(location).updateWithKnownSnapshot(location, snapshot);
+    }
+
+    @Override
+    public FileSystemSnapshot snapshotWithBuilder(Consumer<FileSystemSnapshotBuilder> buildAction) {
+        return buildScopedVirtualFileSystem.snapshotWithBuilder(buildAction);
```

Comment:
Why do this only for the build scoped one?

Reply:
Which file system snapshot should we return when the consumer uses both virtual file systems? This method only allows the caller to build a `FileSystemSnapshot` for "strange" file trees. There is no effect on the virtual file system at all. The only difference would be which hasher the builder uses - and I think the build session scoped hasher is good enough.

Reply:
I didn't realize all we use is the interner and the hasher in that call. Can we inline it instead?

Reply:
I moved `snapshotWithBuilder` completely from snapshotter and virtual file system. The file collection snapshotter needs to do this.